### ISSUE TITLE
Discord emotes aren't turned into text for IRC and certain language-specific characters are sometimes broken

### ISSUE THREAD

**Issue by Kupuntu** [reactiflux/discord-irc#192]
It appears that emotes like :D don't turn from their Discord emoji to text. I don't know if it's the correct way but every time I post an emoji to Discord, it forwards the emoji to IRC and not its text. Is this intended behavior or not?

The other issue is with special language-specific characters like ä, ö and å. This is from IRC to Discord. It works sometimes, as seen here: 
![image](https://cloud.githubusercontent.com/assets/13789597/23085015/dfdd0c68-f56d-11e6-998d-fca53e463e1a.png)

However, sometimes it doesn't: 
![image](https://cloud.githubusercontent.com/assets/13789597/23085021/ee1060dc-f56d-11e6-891f-7ef8ed86ec01.png)

This seems to only happen with certain users in IRC. Whenever they post a message with those characters, it's always broken. For some users it never happens. It could be because others use ISO 8859-1 and others use UTF-8 but I don't know for sure. Is there a way to fix this?

**Comment by Throne3d** [comment]
I'm not sure how to fix the encoding issue – I'd expect it to have problems in individual IRC clients looking at the channel, if the encoding varies between messages, but perhaps they're set up to get around this somehow.

For the former problem, I think the bot presently only translates custom emotes (at least, they must be of the form `<:asdf:1234>`, which I don't think most emotes are). I have no idea if that's intended behavior, I'm afraid.

**Comment by Kupuntu** [comment]
I actually looked into this and it seems that the IRC library could be able to transcode the encoding from ISO 8859-1 to UTF-8 which is understood by Discord.

However what I couldn't find was the emoji conversion to text emotes. Somehow the bot is posting the emoji itself (like this: 😄 this is the direct copy and paste from irssi) instead of something else. I don't know how to parse it to post :D instead.

I might take a look at it at some point when my classes end in a few weeks but I don't know if my skills are good enough. I'm more of a Java programmer and honestly not very good at it either.

**Comment by Throne3d** [comment]
Doesn't look like there's a neat way to parse the emoji into its original text, nor does it look like there's some way to *get* the original text, so I think we'd have to rely on some sort of reverse mapping to translate emojis back into the `:smile:` format – and I can't seem to find anything built into discord.js for doing that how Discord does it.

Maybe I missed something?

**Comment by RheaAyase** [comment]
How about... All matches for regex `:.+:` are replaced by a value from dictionary lookup where key is the match, e.g. `:smile:` and value is some pre-defined replacement. For example:
`:smile:` -> `:D`
`:smiley:` -> `:)`
`:heart:` -> `<3`
etc etc... Just the most commonly used ones? (I would rather avoid unicode emoji)

If the dictionary doesn't contain said value, ignore it, leave it in its raw format, whether it's `:smiling_imp:` or `<:mycustomemoji:123456789>`... Not much else you can do about that...

**Comment by Kupuntu** [comment]
For some reason it seems like the emoticon is never :smile: or anything similar to it. When discord-irc gets the data, it's already a UTF-8 (I think?) emoticon.

In any case, I "fixed" this by turning off the conversion from text to emoticon in the Discord settings. I rarely use any of those outside the IRC anyway.

The ISO 8859-1 charset conversion and detection is still something I'm looking to do.

**Comment by Throne3d** [comment]
@RheaAyase This could be done but seems like it might end up being… out of scope? It seems more like a feature I'd expect `discord.js` to implement instead. It might be worth filing it up there, if you care about it enough?

I'd prefer to use #237 now about the ISO 8859-1 charset thing, to attempt to keep each bug report focused on a single bug. It's probably an upstream bug, [as mentioned there](https://github.com/reactiflux/discord-irc/issues/237#issuecomment-309187525).

**Comment by RheaAyase** [comment]
@Throne3d I don't think so, that's just an api library that gives you a message content. This is the end user who has to process it, if we need it in standard text format `:]` instead of a string to display it reasonably on IRC, it is our problem...

That said, some of the libraries do have two different properties, one for raw text, and one for formatted, where it turns mentions into `@Username` instead of the `<@:id>` etc...

**Comment by Throne3d** [comment]
The library does have such a feature but does not convert emoji into the textual equivalents. I think it's reasonable to expect it to provide a lookup of such textual equivalents, as the Discord desktop client does that when you hover over messages, and I think the library should most of the same information that the desktop client does?

**Comment by Throne3d** [comment]
@tenleftfingers I can try to! (I keep finding myself buried under a bunch of work, so I'm not completely confident.) I hope they're receptive to the idea; the discord desktop app definitely gives ascii human-readable equivalents of these common emoji, if you hover over them, and it would be nice if the library had feature parity in this regard.

**Comment by tenleftfingers** [comment]
Here you go @Throne3d. No pressure! Life comes first. https://github.com/discordjs/discord.js/issues/2898

As I wrote it, I realise it might be a client side issue with no solution on the service side. But I've asked that there anyway. Thank you and I hope work decompresses soon :)

**Comment by Throne3d** [comment]
It looks like the verdict upstream is that they don't think it's in scope for them to provide this, which seems reasonable. I did however find a reference to this node package: <https://www.npmjs.com/package/node-emoji>. It should be relatively straightforward to use the `unemojify` feature of this and add it to the code, perhaps behind a config option, if someone with programming experience wants to be proactive and get to adding this.
